lib/sda/service.go

```diff
@@ -0,0 +1,229 @@
+package sda
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/dedis/cothority/lib/dbg"
+	"github.com/dedis/cothority/lib/network"
+	"github.com/satori/go.uuid"
+	"os"
+	"path"
+)
+
+// Service is a generic interface to define any type of services.
+type Service interface {
+	NewProtocol(*TreeNodeInstance, *GenericConfig) (ProtocolInstance, error)
+	// ProcessRequest is the function that will be called when a external client
+	// using the CLI will contact this service with a request packet.
+	// Each request has a field ServiceID, so each time the Host (dispatcher)
+	// receives a request, it looks whether it knows the Service it is for and
+	// then dispatch it through ProcessRequest.
+	ProcessRequest(*network.Entity, *Request)
+	// XXX Later we could introduce:
+	// ProcessServiceRequest(*network.Entity,  *Request)
+}
+
+type ServiceID uuid.UUID
+
+func (s *ServiceID) String() string {
+	return uuid.UUID(*s).String()
+}
+
+var NilServiceID = ServiceID(uuid.Nil)
+
+// NewServiceFunc is the type of a function that is used to instantiate a given Service
+// A service is initialized with a Host (to send messages to someone), the
+// overlay (to register a Tree + EntityList + start new node), and a path where
+// it can finds / write everything it needs
+type NewServiceFunc func(c Context, path string) Service
+
+// GenericConfig is a config that can withhold any type of specific configs for
+// protocols. It is passed down to the service NewProtocol function.
+type GenericConfig struct {
+	Type uuid.UUID
+	//Data network.ProtocolMessage
+}
+
+var GenericConfigID = network.RegisterMessageType(GenericConfig{})
+
+// A serviceFactory is used to register a NewServiceFunc
+type serviceFactory struct {
+	cons map[ServiceID]NewServiceFunc
```

Comment:
I think @nikkolasg wrote `cons` because it is also used in protobuf (ans d I think I saw it in other places too). My personal rule of thumb is: "long" (speaking) names for everything that is exported. For local vars used only in a short scope I use very short var names. For unexported vars/fields (like here), I also tend to use a tradeoff between the two. But clear, "speaking" names never hurt (only while typing ;-))